src/wal.rs

```diff
@@ -1,10 +1,60 @@
 use super::Result;
+use crate::util::{decode_varint_uncheck, encode_varint_uncheck, varint_len};
+use bytes::{BufMut, Bytes, BytesMut};
 use std::fs::{File, OpenOptions};
 use std::path::PathBuf;
 
+#[derive(Default, Debug, PartialEq)]
 struct Header {
-    key_len: usize,
-    value_len: usize,
+    key_len: u32,
+    value_len: u32,
+    expires_at: u64,
+    meta: u8,
+    user_meta: u8,
+}
+
+impl Header {
+    pub fn encoded_len(&self) -> usize {
+        1 + 1
+            + varint_len(self.expires_at as usize)
+            + varint_len(self.key_len as usize)
+            + varint_len(self.value_len as usize)
+    }
+
+    pub fn encode(&self, bytes: &mut BytesMut) {
+        let encoded_len = self.encoded_len();
+        bytes.reserve(encoded_len);
+        let read = unsafe {
+            let buf = bytes.bytes_mut();
```

Comment:
Perhaps adding an if check before entering `get_unchecked_mut` is better.

Reply:
I've added the check `assert!(buf.len() >= encoded_len);`. Seems that `reserve` will always ensure the capacity is enough, I'm not sure whether this length check is necessary or not.